### PR TITLE
852 adds a timeout call decorator

### LINKED ISSUE #852: unit tests with expected execution time
**Is your feature request related to a problem? Please describe.**
in many cases it is necessary to check whether a unit test could finish within some expected execution time, in addition to the correctness checks of the outputs.

**Describe the solution you'd like**
this ticket looks for a generic timer wrapper for the unit testing module, so that it automatically fails a test (early) if the test takes longer than the expected execution time when running it on the monai CI/CD servers.

would also be great to have a global flag to switch this feature on/off so that the developers can run the tests locally without an execution time constraint

### LINKED ISSUE #927: Add execute-time test for integration test cases
**Is your feature request related to a problem? Please describe.**
We test the accuracy with determinism training in integration tests, need to also test the max execute-time.
Then we may also add this time test to many unit tests.

### PR DESCRIPTION
**Pull request by wyli**
Signed-off-by: Wenqi Li <wenqil@nvidia.com>

fixes #852 fixes #927 

### Description
adds a test util for raising an exception if the execution time exceeds a threshold

### Status
**Ready**

### Types of changes

- [x] Non-breaking change (fix or new feature that would not break existing functionality).
- [x] New tests added to cover the changes.
- [x] Integration tests passed locally by running `./runtests.sh --codeformat --coverage`.
- [x] Quick tests passed locally by running `./runtests.sh --quick`.
- [x] In-line docstrings updated.
- [x] Documentation updated, tested `make html` command in the `docs/` folder.

### CODE REVIEW

**Comment by Nic-Ma** [comment]
Hi @IsaacYangSLA ,

Could you please help review this PR to enhance our CICD?

Thanks.

**Comment by wyli** [comment]
Hi @Nic-Ma could you review it as well, this PR also added some testing logic for the integration tests you created earlier

**Comment by Nic-Ma** [comment]
OK, sure

File: tests/test_integration_classification_2d.py [hunk 1]
```diff
@@ -240,6 +240,33 @@ def test_training(self):
 
         np.testing.assert_allclose(repeated[0], repeated[1])
 
+    @TimedCall(seconds=400, skip_timing=not torch.cuda.is_available(), daemon=False)
+    def test_timing(self):
```
Comment (Nic-Ma) [inline tests/test_integration_classification_2d.py:244]: I think most of the logic in this function is duplicated to `test_training`, could you please combine them or extract a common internal function?

File: tests/test_integration_segmentation_3d.py [hunk 2]
```diff
@@ -282,6 +282,28 @@ def test_training(self):
         np.testing.assert_allclose(repeated[0], repeated[1])
         np.testing.assert_allclose(repeated[0], repeated[2])
 
+    @TimedCall(seconds=180, daemon=False)
+    def test_timing(self):
```
Comment (Nic-Ma) [inline tests/test_integration_segmentation_3d.py:286]: I think most of the logic in this function is duplicated to test_training, could you please combine them or extract a common internal function?

File: tests/test_integration_workflows.py [hunk 3]
```diff
@@ -288,6 +288,20 @@ def test_training(self):
                 self.assertTrue(test_integration_value(TASK, key="output_sums", data=repeated[i][2:], rtol=1e-2))
         np.testing.assert_allclose(repeated[0], repeated[1])
 
+    @TimedCall(seconds=200, skip_timing=not torch.cuda.is_available(), daemon=False)
+    def test_timing(self):
```
Comment (Nic-Ma) [inline tests/test_integration_workflows.py:292]: I think most of the logic in this function is duplicated to test_training, could you please combine them or extract a common internal function?

**Review by Nic-Ma — APPROVED** [review]
This overall feature looks good to me except some minor comments.
@IsaacYangSLA Is it expected in our CI / CD development plan for the test of execution time?
If you don't have any more comments, we will merge it later.

Thanks.

**Pull request merged**